src/p2p/swarm.rs

```diff
@@ -108,15 +148,26 @@ impl SwarmApi {
     pub fn connect(&mut self, target: ConnectionTarget) -> SubscriptionFuture<Result<(), String>> {
         trace!("Connecting to {:?}", target);
 
-        self.events.push_back(match target {
-            ConnectionTarget::Addr(ref addr) => NetworkBehaviourAction::DialAddress {
-                address: addr.clone(),
-            },
-            ConnectionTarget::PeerId(ref id) => NetworkBehaviourAction::DialPeer {
-                peer_id: id.clone(),
-                condition: DialPeerCondition::Disconnected,
-            },
-        });
+        match target {
+            ConnectionTarget::Addr(ref addr) => {
+                self.events.push_back(NetworkBehaviourAction::DialAddress {
+                    address: addr.clone(),
+                })
+            }
+            ConnectionTarget::PeerId(ref id) => {
+                self.events.push_back(NetworkBehaviourAction::DialPeer {
+                    peer_id: id.clone(),
+                    condition: DialPeerCondition::Disconnected,
+                })
+            }
+            ConnectionTarget::PeerWithAddrs(ref _peer_id, ref addrs) => {
+                for addr in addrs {
+                    self.events.push_back(NetworkBehaviourAction::DialAddress {
+                        address: addr.clone(),
+                    })
```

Comment:
I wonder if the addresses could just be stored for the PeerId in some swarmapi variable and these should always yield just DialPeer, with DialPeerCondition::Disconnected?

Reply:
Hmm, but doesn't this assume that dialing the `PeerId` is preferable? This puts plenty of trust in the DHT; maybe I could do that and put a timeout on that attempt, so that the addresses can be used as a fallback method?

Reply:
I think dialing the PeerId is always preferable.

> This puts plenty of trust in the DHT

I was not suggesting changing kademlia state; SwarmAPI can provide the addresses through it's NetworkBehaviour implementation: https://github.com/rs-ipfs/rust-ipfs/blob/b099deb4aa9d0dcee9d140d0c1f203941ff4fd37/src/p2p/swarm.rs#L212-L218.

> maybe I could do that and put a timeout on that attempt, so that the addresses can be used as a fallback method?

Just to clarify DHT should be used when user uses `/api/v0/swarm/connect?arg=/p2p/peer_id` but we don't support that at the moment at all. Whenever a req like `/api/v0/swarm/connect?arg=/.../p2p/peer_id` comes in, we should use the provided addresses. This was done in a bit of haste back then so it might not fit well into the current implementation.

I however do not know how does this work with *kademlia* knowing a address for this peer. Should kademlia addresses always be attempted last, so that users provided addresses are preferred? Quite likely. Without looking at the NetworkBehaviour derive impl, I would guess that this ordering is "specified" by ordering kademlia as the last field.

Reply:
> DHT should be used when user uses /api/v0/swarm/connect?arg=/p2p/peer_id

Ok, all clear now; I just thought this might have been used in the `/api/v0/swarm/connect?arg=/.../p2p/peer_id` cases instead.